File: elixir-shyryan/lib/lexer.ex
defmodule Monkey.Lexer do
  @moduledoc """
  The Monkey Lang lexer in Elixir.
  """

  alias Monkey.Token

  defguardp is_whitespace(c) when c in ~c[ \n\t]
  defguardp is_letter(c) when c in ?a..?z or c in ?A..?Z or c == ?_
  defguardp is_digit(c) when c in ?0..?9

  @doc """
  Turn some input into a list of tokens.

  ## Example

      iex> Lexer.init("let five = 5;")
      [
        %Token{type: :let, literal: "let"},
        %Token{type: :ident, literal: "five"},
        %Token{type: :assign, literal: "="},
        %Token{type: :int, literal: "5"},
        %Token{type: :semicolon, literal: ";"},
        %Token{type: :eof, literal: ""}
      ]

  """
  @spec init(String.t()) :: [Token.t()]
  def init(input) when is_binary(input) do
    tokenize(input, [])
  end

  # Recursive base-case. When the input is empty, then we add an EOF token.
  defp tokenize(<<>>, tokens) do
    [Token.new(:eof) | tokens] |> Enum.reverse()
  end

  # Ignore whitespace.
  defp tokenize(<<c::8, rest::binary>>, tokens) when is_whitespace(c),
    do: tokenize(rest, tokens)

  # Recursively go through the input, tokenizing each character.
  # Uses binary pattern-matching to match on the first character(s).
  # See: https://hexdocs.pm/elixir/Kernel.SpecialForms.html#<<>>/1
  defp tokenize(input, tokens) do
    {token, rest} =
      case input do
        <<c::8, _::binary>> when is_letter(c) -> read_identifier(input)
        <<c::8, _::binary>> when is_digit(c) -> read_number(input)
        <<"==", rest::binary>> -> {Token.new(:equal_equal), rest}
        <<"!=", rest::binary>> -> {Token.new(:not_equal), rest}
        <<";", rest::binary>> -> {Token.new(:semicolon), rest}
        <<",", rest::binary>> -> {Token.new(:comma), rest}
        <<"(", rest::binary>> -> {Token.new(:lparen), rest}
        <<")", rest::binary>> -> {Token.new(:rparen), rest}
        <<"{", rest::binary>> -> {Token.new(:lsquirly), rest}
        <<"}", rest::binary>> -> {Token.new(:rsquirly), rest}
        <<"=", rest::binary>> -> {Token.new(:assign), rest}
        <<"+", rest::binary>> -> {Token.new(:plus), rest}
        <<"-", rest::binary>> -> {Token.new(:minus), rest}
        <<"!", rest::binary>> -> {Token.new(:bang), rest}
        <<"/", rest::binary>> -> {Token.new(:slash), rest}
        <<"*", rest::binary>> -> {Token.new(:asterisk), rest}
        <<">", rest::binary>> -> {Token.new(:greater_than), rest}
        <<"<", rest::binary>> -> {Token.new(:less_than), rest}
        <<"return", rest::binary>> -> {Token.new(:return), rest}
        <<_::8, rest::binary>> -> {Token.new(:illegal), rest}
      end

    tokenize(rest, [token | tokens])
  end

  defp read_identifier(input) do
    {identifier, rest} = read_identifier(input, "")

    token =
      identifier
      |> Token.lookup_identifier()
      |> Token.new(identifier)

    {token, rest}
  end

  defp read_identifier(<<c::8, rest::binary>>, acc) when is_letter(c) do
    read_identifier(rest, acc <> <<c>>)
  end

  defp read_identifier(rest, identifier), do: {identifier, rest}

  defp read_number(input) do
    {number, rest} = read_number(input, "")
    {Token.new(:int, number), rest}
  end

  defp read_number(<<c::8, rest::binary>>, acc) when is_digit(c) do
    read_number(rest, acc <> <<c>>)
  end

  defp read_number(rest, number), do: {number, rest}
end

File: typescript-types2/src/lexer/__tests__/lexer.ts
import { GetNextToken, GetToken, Token, TokenType, Tokenizer } from "..";

export type Assert<T extends true> = T;
export type Equal<X, Y> = (<T>() => T extends X ? 1 : 2) extends <
  T
>() => T extends Y ? 1 : 2
  ? true
  : false;

/**
 * check that the first {@code TokenTypes.length} tokens' types from {@code Lexer} match the passed {@code TokenTypes}.
 *
 * @template Lexer a Lexer instance
 * @template TokenTypes the token types to match
 */
export type CheckTokenTypes<
  T extends Tokenizer,
  TokenTypes extends TokenType[]
> = TokenTypes["length"] extends 0
  ? true
  : GetNextToken<T> extends infer T extends Tokenizer
  ? Equal<GetToken<T>["type"], TokenTypes[0]> extends true
    ? CheckTokenTypes<
        T,
        TokenTypes extends [TokenType, ...infer Rest extends TokenType[]]
          ? Rest
          : []
      >
    : {
        expected: TokenTypes[0];
        got: GetToken<T>;
      }
  : "next token not a Tokenizer";

/**
 * check that the first {@code Tokens.length} tokens from {@code Lexer} match the passed {@code Tokens}.
 *
 * @template Lexer a Lexer instance
 * @template Tokens the tokens to match
 */
export type CheckTokens<
  Lexer extends Tokenizer,
  Tokens extends Token[]
> = Tokens["length"] extends 0
  ? true
  : GetNextToken<Lexer> extends infer T extends Tokenizer
  ? Equal<GetToken<T>, Tokens[0]> extends true
    ? CheckTokens<
        T,
        Tokens extends [Token, ...infer Rest extends Token[]] ? Rest : []
      >
    : {
        expected: Tokens[0];
        got: GetToken<T>;
      }
  : "Next token not a Tokenizer";

/** test GetNextToken */
export namespace TestGetNextTokenToken {
  type Input = `=+(){},;`;
  export type Lexer = Tokenizer<Input>;

  type Tokens = [
    TokenType.Assign,
    TokenType.Plus,
    TokenType.LParen,
    TokenType.RParen,
    TokenType.LSquirly,
    TokenType.RSquirly,
    TokenType.Comma,
    TokenType.Semicolon
  ];

  export type Result = Assert<CheckTokenTypes<Lexer, Tokens>>;
}

/** test GetNextToken complete */
export namespace TestGetNextTokenComplete {
  type Input = `let five = 5;
        let ten = 10;
        let add = fn(x, y) {
            x + y;
        };
        let result = add(five, ten);
        !-/*5;
        5 < 10 > 5;
        if (5 < 10) {
            return true;
        } else {
            return false;
        }

        10 == 10;
        10 != 9;
        `;
  type Lexer = Tokenizer<Input>;

  type Tokens = [
    { type: TokenType.Let; literal: "let" },
    { type: TokenType.Ident; literal: "five" },
    { type: TokenType.Assign; literal: "=" },
    { type: TokenType.Int; literal: "5" },
    { type: TokenType.Semicolon; literal: ";" },
    { type: TokenType.Let; literal: "let" },
    { type: TokenType.Ident; literal: "ten" },
    { type: TokenType.Assign; literal: "=" },
    { type: TokenType.Int; literal: "10" },
    { type: TokenType.Semicolon; literal: ";" },
    { type: TokenType.Let; literal: "let" },
    { type: TokenType.Ident; literal: "add" },
    { type: TokenType.Assign; literal: "=" },
    { type: TokenType.Function; literal: "fn" },
    { type: TokenType.LParen; literal: "(" },
    { type: TokenType.Ident; literal: "x" },
    { type: TokenType.Comma; literal: "," },
    { type: TokenType.Ident; literal: "y" },
    { type: TokenType.RParen; literal: ")" },
    { type: TokenType.LSquirly; literal: "{" },
    { type: TokenType.Ident; literal: "x" },
    { type: TokenType.Plus; literal: "+" },
    { type: TokenType.Ident; literal: "y" },
    { type: TokenType.Semicolon; literal: ";" },
    { type: TokenType.RSquirly; literal: "}" },
    { type: TokenType.Semicolon; literal: ";" },
    { type: TokenType.Let; literal: "let" },
    { type: TokenType.Ident; literal: "result" },
    { type: TokenType.Assign; literal: "=" },
    { type: TokenType.Ident; literal: "add" },
    { type: TokenType.LParen; literal: "(" },
    { type: TokenType.Ident; literal: "five" },
    { type: TokenType.Comma; literal: "," },
    { type: TokenType.Ident; literal: "ten" },
    { type: TokenType.RParen; literal: ")" },
    { type: TokenType.Semicolon; literal: ";" },

    { type: TokenType.Bang; literal: "!" },
    { type: TokenType.Dash; literal: "-" },
    { type: TokenType.ForwardSlash; literal: "/" },
    { type: TokenType.Asterisk; literal: "*" },
    { type: TokenType.Int; literal: "5" },
    { type: TokenType.Semicolon; literal: ";" },
    { type: TokenType.Int; literal: "5" },
    { type: TokenType.LessThan; literal: "<" },
    { type: TokenType.Int; literal: "10" },
    { type: TokenType.GreaterThan; literal: ">" },
    { type: TokenType.Int; literal: "5" },
    { type: TokenType.Semicolon; literal: ";" },

    { type: TokenType.If; literal: "if" },
    { type: TokenType.LParen; literal: "(" },
    { type: TokenType.Int; literal: "5" },
    { type: TokenType.LessThan; literal: "<" },
    { type: TokenType.Int; literal: "10" },
    { type: TokenType.RParen; literal: ")" },
    { type: TokenType.LSquirly; literal: "{" },
    { type: TokenType.Return; literal: "return" },
    { type: TokenType.True; literal: "true" },
    { type: TokenType.Semicolon; literal: ";" },
    { type: TokenType.RSquirly; literal: "}" },
    { type: TokenType.Else; literal: "else" },
    { type: TokenType.LSquirly; literal: "{" },
    { type: TokenType.Return; literal: "return" },
    { type: TokenType.False; literal: "false" },
    { type: TokenType.Semicolon; literal: ";" },
    { type: TokenType.RSquirly; literal: "}" },

    { type: TokenType.Int; literal: "10" },
    { type: TokenType.Equal; literal: "==" },
    { type: TokenType.Int; literal: "10" },
    { type: TokenType.Semicolon; literal: ";" },
    { type: TokenType.Int; literal: "10" },
    { type: TokenType.NotEqual; literal: "!=" },
    { type: TokenType.Int; literal: "9" },
    { type: TokenType.Semicolon; literal: ";" },

    { type: TokenType.Eof; literal: "\0" }
  ];
  export type Result = Assert<CheckTokens<Lexer, Tokens>>;
}
